src/typescript-worker/types-fetcher.ts

```diff
@@ -0,0 +1,291 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import * as ts from 'typescript';
+import {ModuleResolver, isRelativeOrAbsolutePath} from './module-resolver.js';
+import {Deferred} from '../shared/deferred.js';
+import {Result} from '../shared/util.js';
+
+/**
+ * Fetches typings for TypeScript imports and their transitive dependencies, and
+ * for standard libraries.
+ */
+export class TypesFetcher {
+  private readonly _moduleResolver: ModuleResolver;
+  private readonly _entrypointTasks: Promise<void>[] = [];
+  private readonly _handledSpecifiers = new Set<string>();
+  private readonly _specifierToFetchResult = new Map<
+    string,
+    Deferred<Result<string, number>>
+  >();
+
+  constructor(moduleResolver: ModuleResolver) {
+    this._moduleResolver = moduleResolver;
+  }
+
+  /**
+   * Start fetching type definitions for all bare module specifiers in the given
+   * TypeScript source text. Relative module specifiers are ignored.
```

Comment:
>Relative module specifiers are ignored

relative specifiers are followed for transitive imports, yes? A dependency imported via a bare specifier might relatively import a same-package module that has typings.

Reply:
Yep, relative specifiers are supported, just not from the top level. We don't care about them in the top level, because any relative path there would be a project file, which we are already handling. But if a project file imports a bare module, which imports a relative path, it will be handled correctly.